src/nvmet.c

```diff
@@ -833,209 +909,230 @@ int nvmet_revalidate_size(struct RD_PROFILE *rd_prof, RC_PROFILE *rc_prof, char
+		rc = INVALID_VALUE;
+
+		free_nvmet_linked_lists(NULL, orig_profile);
 	}
 
-	/*
-	 * Grab number of entries in allowed_hosts directory. We need this later, to check if one of them has been removed.
+
+	/* Example of this path:
+	 * /sys/kernel/config/nvmet/subsystems/nqn.2021-06.org.rapiddisk:ubuserver-rd0/allowed_hosts
 	 */
-	sprintf(path, "%s/%s%s-%s/allowed_hosts/", SYS_NVMET_TGT, NQN_HDR_STR, hostname, device);
+	sprintf(path, "%s/%s/allowed_hosts", SYS_NVMET_TGT, nqn);
 	if ((allowed_host_number = scandir(path, &list, scandir_filter_no_dot, NULL)) < 0) {
-		msg = "Error. Unable to scan host exports directory for %s. %s: unlink: %s";
+		msg = "Error. Unable to access %s. %s: scandir: %s";
 		print_error(msg, return_message, path, __func__, strerror(errno));
 		return rc;
 	}
+
+	if (allowed_host_number > 0) {
+		for (n = 0; n < allowed_host_number; n++) {
+			strcat(target_string, list[n]->d_name);
+			strcat(target_string, " ");
+		}
+		size_t size = strlen(target_string);
+		size = size - 1;
+		target_string[size] = '\0';
+	}
+
 	list = clean_scandir(list, allowed_host_number);
 
-	/* Remove host NQN */
-	if (strlen(host) != 0) {
-		sprintf(path, "%s/%s%s-%s/allowed_hosts/%s", SYS_NVMET_TGT, NQN_HDR_STR, hostname, device, host);
+	if (allowed_host_number > 0 && strlen(host) <= 0) {
+		msg = "Error. No host provided, but one or more target(s) is/are still mapped to NQN(s): %s. Please remove them first.";
+		print_error(msg, return_message, target_string);
+		return rc;
+	} else if (allowed_host_number > 0 && strlen(host) > 0) {
+		/* Example of this path:
+	 	 * /sys/kernel/config/nvmet/subsystems/nqn.2021-06.org.rapiddisk:ubuserver-rd0/allowed_hosts/<HOST>
+	 	 */
+		sprintf(path, "%s/%s/allowed_hosts/%s", SYS_NVMET_TGT, nqn, host);
 		if (access(path, F_OK) == SUCCESS) {
-			rc = unlink(path);
-			if (rc != SUCCESS) {
-				msg = "Error. Unable to remove host. %s: unlink: %s";
-				print_error(msg, return_message, __func__, strerror(errno));
+			if (unlink(path) != SUCCESS) {
+				msg = "Error. Unable to remove link %s. %s: unlink: %s";
+				print_error(msg, return_message, path, __func__, strerror(errno));
 				return rc;
 			}
-			msg = "Block device %s has been unmapped from NVMe Initiator host %s.";
-			if (return_message == NULL) {
-				print_error(msg, return_message, device, host);
-			} else {
-				sprintf(json_message_temp, msg, device, host);
-				strcat(json_message, json_message_temp);
+		}
+		/* Example of this path:
+	 	 * /sys/kernel/config/nvmet/hosts/<HOST>
+	 	 */
+		sprintf(path, "%s/%s", SYS_NVMET_HOSTS, host);
+		if (access(path, F_OK) == SUCCESS) {
+			if (rmdir(path) != SUCCESS) {
+				msg = "Error. Unable to remove dir %s. %s: rmdir: %s";
+				print_error(msg, return_message, path, __func__, strerror(errno));
+				return rc;
 			}
-		} else {
-			msg = "NVMe Initiator Host '%s' does not exist.";
-			print_error(msg, return_message, host);
-			return INVALID_VALUE;
+			allowed_host_number--;
```

Comment:
@matteotenca Shouldn't this line:
allowed_host_number--;
go to up where the unlink is above?

Reply:
Yes, you're right, I moved it in correct place.